src/comm/XPUPair.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <ATen/Context.h>
+
+namespace at::xpu {
+
+template <typename T1, typename T2>
+struct pair {
+  typedef T1 first_type;
+  typedef T2 second_type;
+
+  first_type first;
+  second_type second;
+
+  // default constructor
+  pair(void) : first(), second() {}
+
+  inline pair(const T1& x, const T2& y) : first(x), second(y) {}
+
+  template <typename U1, typename U2>
+  inline pair(const pair<U1, U2>& p) : first(p.first), second(p.second) {}
+
+  template <typename U1, typename U2>
+  pair(const std::pair<U1, U2>& p) : first(p.first), second(p.second) {}
+
+  inline void swap(pair& p) {
+    std::swap(first, p.first);
```

Comment:
This still doesn't work with custom swap (e.g. if you have `pair<pair<...>...>`). Correct use needs to enable ADL. For details see e.g. https://stackoverflow.com/questions/28130671/how-does-using-stdswap-enable-argument-dependent-lookup-adl.

Nit: This member function seems unnecessary. The implementation could go directly into the free-function.

Reply:
@xytintel Please address.

Reply:
Removed